### PR TITLE
ISSUE 2243 - Metadata rule query APIs

### LINKED ISSUE #2243: Rules for Metadata
### Description

Use rule queries from groups to query metadata

### Tasks

- [x] Refactor rules out of groups
- [x] Refactor rules to helper/rules
- [x] Refactor metadata out of helper/model
- [x] Add tests for federation smart group
- [x] Add tests for meta filters with rules
- [x] Get list of mesh ID based on smart rules
- [x] Add tests for filter all metadata with rules
- [x] All metadata filter based on smart rules
- [x] Remove mongoose dep?
- [ ] Clean up

### PR DESCRIPTION
**Pull request by Charence**
This fixes #2243

#### Description
- Support for mesh ID query with rules
- Support for metadata query with rules

#### Test cases
- Mesh ID query
```
POST /:teamspace/:model/revision/:revisionId/meta/rules?meshids=true
[
    {
        "field": "Area",
        "operator": "GT",
        "values": [ 10 ]
    }
]
```
- Metadata query
```
POST /:teamspace/:model/revision/:revisionId/meta/rules
[
    {
        "field": "Area",
        "operator": "GT",
        "values": [ 10 ]
    }
]
```

### CODE REVIEW

File: backend/routes/meta.js [hunk 1]
```diff
@@ -273,14 +376,15 @@
...
 		let branch;
 
 		if (!req.params.rev) {
 			branch = C.MASTER_BRANCH_NAME;
 		}
 
-		ModelHelpers.getAllIdsWith4DSequenceTag(req.params.account, req.params.model, branch, req.params.rev)
+		Meta.getMeshIdsByRules(req.params.account, req.params.model, branch, req.params.rev, rules)
 			.then(obj => {
 				responseCodes.respond(utils.APIInfo(req), req, res, next, responseCodes.OK, obj, undefined, req.param.rev ? config.cachePolicy : undefined);
```
Comment (carmenfan) [inline backend/routes/meta.js:389]: can't define cache policy here you'd end up caching previous results which could've been a completely different rule.

File: backend/routes/meta.js [hunk 2]
```diff
@@ -289,14 +393,15 @@
...
 		let branch;
 
 		if (!req.params.rev) {
 			branch = C.MASTER_BRANCH_NAME;
 		}
 
-		ModelHelpers.getAllIdsWithMetadataField(req.params.account, req.params.model, branch, req.params.rev, req.params.metaKey)
+		Meta.getAllMetadataByRules(req.params.account, req.params.model, branch, req.params.rev, rules)
 			.then(obj => {
 				responseCodes.respond(utils.APIInfo(req), req, res, next, responseCodes.OK, obj, undefined, req.param.rev ? config.cachePolicy : undefined);
```
Comment (carmenfan) [inline backend/routes/meta.js:406]: same here

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: backend/package.json [hunk 3]
```diff
@@ -15,6 +15,7 @@
     "test:one": "cross-env NODE_ENV=test NODE_CONFIG_DIR=../config mocha",
     "test:int-clean": "./test/cleanup/cleanup.sh",
     "test:int": "cross-env NODE_ENV=test NODE_CONFIG_DIR=../config mocha test/integrated",
+    "test:int-one": "cross-env NODE_ENV=test NODE_CONFIG_DIR=../config mocha ",
```
[the remark this answers is not in the document]
Comment (Charence) [reply backend/package.json:18]: @carmenfan I didn't see that 😅

File: backend/routes/meta.js [hunk 4]
```diff
@@ -273,14 +361,43 @@
...
+		const rules = req.body;
+		let branch;
+
+		if (!req.params.rev) {
+			branch = C.MASTER_BRANCH_NAME;
+		}
+
+		const showMeshIds = (req.query.meshids) ? JSON.parse(req.query.meshids) : false;
+
+		if (showMeshIds) {
```
Comment (carmenfan) [inline backend/routes/meta.js:374]: less duplication if we do this 
```suggestion
                const promise = showMeshIds ? Meta.getMeshIdsByRules(req.params.account, req.params.model, branch, req.params.rev, rules) : Meta.getAllMetadataByRules(req.params.account, req.params.model, branch, req.params.rev, rules);
                
                promise.then((obj) => ...)

```

File: backend/routes/meta.js [hunk 5]
```diff
@@ -273,14 +361,43 @@
...
+			Meta.getMeshIdsByRules(req.params.account, req.params.model, branch, req.params.rev, rules)
+				.then(obj => {
+					responseCodes.respond(utils.APIInfo(req), req, res, next, responseCodes.OK, obj, undefined, undefined);
```
Comment (carmenfan) [inline backend/routes/meta.js:377]: you don't need to pass undefineds if they're at the end of the function calls. by the virtue of not passing over something they'd be undefined.
```suggestion
					responseCodes.respond(utils.APIInfo(req), req, res, next, responseCodes.OK, obj);
```

**Review by carmenfan — CHANGES_REQUESTED** [review]

**Comment by carmenfan** [comment]
@3drepo/product-team  if you can confirm this does what you wanted http://issue-2243.dev.3drepo.io/

File: backend/routes/meta.js [hunk 6]
```diff
@@ -119,18 +118,107 @@
...
+	 *
+	 * @apiExample {post} Example usage (/:revId)
+	 * POST /teamSpace1/3549ddf6-885d-4977-87f1-eeac43a0e818/revision/00000000-0000-0000-0000-000000000001/meta/rules HTTP/1.1
+	 *
+	 * @apiExample {post} Example usage (mesh IDs)
+	 * POST /teamSpace1/3549ddf6-885d-4977-87f1-eeac43a0e818/revision/master/head/meta/rules?meshids=true HTTP/1.1
+	 *
+	 * @apiSuccessExample {json} Success (metadata):
+	 * {
+	 *    "data": [
```
Comment (carmenfan) [inline backend/routes/meta.js:154]: I think we may need to change this return to tell the user which model it came from... If this endpoint points to a federation, the metadata are always in the submodels, and at the moment there's no way to tell where it came from with this schema.

thoughts @mosment3DRepo ?

**Review by mosment3DRepo — COMMENTED** [review]
I'm having a problem with the API requests, and I'm not sure if it's me or the API

POST https://api1.issue-2243.dev.3drepo.io/api/carmen/067a7522-6dd8-4258-97a0-be756e596324/revision/41654b92-0255-4f69-ad2b-c33f381c99d0/meta/rules

with this body:
>[
    {
        "field": "IFC GUID",
        "operator": "IS",
        "values": [ "3jMXdawx93suHqay4NiXc7" ]
    }
]

Returns the following:

> {
    "data": [
        {
            "_id": "f8902eac-35f1-4cde-b80f-2536cef559d7",
            "parents": [
                "04171de0-a1ab-400c-9dba-b0c2d60c954c"
            ],
            "metadata": {
                "IFC Type": "IfcBeam",
                "IFC GUID": "3jMXdawx93suHqay4NiXc7",
                "Reference Level": "Second Floor",
                "Start Level Offset": -0.102,
                "End Level Offset": -0.102,
                "z-Direction Justification": 0,
                "z-Direction Offset Value": 0,
                "Lateral Justification": 0,
                "Cross-Section Rotation": 0,
                "Start Release": 3,
                "Start Fx": "False",
                "Start Fy": "False",
                "Start Fz": "False",
                "Start Mx": "False",
                "Start My": "True",
                "Start Mz": "True",
                "End Release": 1,
                "End Fx": "False",
                "End Fy": "False",
                "End Fz": "False",
                "End Mx": "True",
                "End My": "True",
                "End Mz": "True",
                "Analyze As": 1,
                "Phase Created": "New Construction",
                "Stick Symbol Location": 0,
                "Cut Length": 2.1071,
                "Structural Usage": 6,
                "Start Attachment Type": 4,
                "Rigid Links": 0,
                "Horizontal Projection": "Default",
                "Length": 2.28,
                "Volume": 0.015797,
                "Start Extension": -0.0127,
                "End Extension": -0.0127,
                "Start Extension Calculation": 3,
                "End Extension Calculation": 3,
                "InstallationDate": "InstallationDate",
                "SerialNumber": "SerialNumber",
                "WarrantyStartDate": "WarrantyStartDate",
                "BarCode": "BarCode",
                "AssetIdentifier": "AssetIdentifier",
                "TagNumber": "TagNumber",
                "Reference": "Reference",
                "IsExternal": "False",
                "LoadBearing": "True",
                "Span": 2.1071,
                "Slope": 0,
                "Beam Material": "Metal - Steel - 345 MPa",
                "Manufacturer": "Manufacturer",
                "OmniClass Number": "23.25.30.11.14.14",
                "OmniClass Title": "Beams",
                "bf": 0.153,
                "d": 0.455,
                "k": 0.031,
                "kr": 0.0177,
                "tf": 0.0133,
                "tw": 0.008,
                "AccessibilityPerformance": "AccessibilityPerformance",
                "CodePerformance": "CodePerformance",
                "Color": "Color",
                "Constituents": "Constituents",
                "Features": "Features",
                "Finish": "Finish",
                "Grade": "Grade",
                "Material": "Material",
                "ModelReference": "ModelReference",
                "NominalHeight": "NominalHeight",
                "NominalLength": "NominalLength",
                "NominalWidth": "NominalWidth",
                "ProductionYear": "ProductionYear",
                "Shape": "Shape",
                "Size": "Size",
                "SustainabilityPerformance": "SustainabilityPerformance",
                "WarrantyDescription": "WarrantyDescription",
                "WarrantyDurationLabor": "WarrantyDurationLabor",
                "WarrantyDurationParts": "WarrantyDurationParts",
                "WarrantyGuarantorLabor": "WarrantyGuarantorLabor",
                "WarrantyGuarantorParts": "WarrantyGuarantorParts",
                "ExpectedLife": "ExpectedLife",
                "ReplacementCost": "ReplacementCost",
                "AssetAccountingType": "FIXED",
                "ModelLabel": "ModelLabel",
                "W": 0.58,
                "A": 0.00759
            }
        },
        {
            "_id": "95ab3708-6b16-450a-9d41-40dcb1c3b70a",
            "parents": [
                "60769d87-d362-4b67-9eb7-8426cf72792d"
            ],
            "metadata": {
                "IFC Type": "IfcBeam",
                "IFC GUID": "3jMXdawx93suHqay4NiXc7",
                "Reference": "Reference",
                "LoadBearing": "True",
                "IsExternal": "False",
                "Slope": 0,
                "Span": 2.28,
                "ModelReference": "ModelReference",
                "ModelLabel": "ModelLabel",
                "Manufacturer": "Manufacturer",
                "ProductionYear": "ProductionYear"
            }
        }
    ]
}

When a GET request is performed here: [https://api1.issue-2243.dev.3drepo.io/api/carmen/067a7522-6dd8-4258-97a0-be756e596324/revision/41654b92-0255-4f69-ad2b-c33f381c99d0/meta/all.json](https://api1.issue-2243.dev.3drepo.io/api/carmen/067a7522-6dd8-4258-97a0-be756e596324/revision/41654b92-0255-4f69-ad2b-c33f381c99d0/meta/all.json)

There is only one object with an IFC GUID of 3jMXdawx93suHqay4NiXc7

File: backend/routes/meta.js (hunk 6 above)
Comment (Charence) [reply backend/routes/meta.js:154]: @carmenfan @mosment3DRepo At the moment, it is supposed to match the response for getAllMetadata. For federations, it should return information about the submodel. Maybe I should make this clearer in the documentation.

A federation response will look a bit like this:
```
{
    "data": [],
    "subModels": [
        {
            "data": [
                {
                    "_id": "01bbf2d0-a9a1-4593-b8a7-e185a791ffdb",
                    "parents": [
                        "73fe5634-fb56-481d-95f8-3666bcc65c17"
                    ],
                    "metadata": {
                        "IFC Type": "IfcBuildingElementProxy",
                        "IFC GUID": "1HOEJRLv561Qi4FACVNHPH",
                        "Reference": "Family2",
                        "Host": "Level : Level 1",
                        "Level": "Level: Level 1",
                        "Moves With Nearby Elements": "False",
                        "Offset": 1432.044264,
                        "Area": 2767.039224,
                        "Volume": 4504.122208,
                        "Category": "Site",
                        "Family": "Family2: Family2",
                        "Family and Type": "Family2: Family2",
                        "Type": "Family2: Family2",
                        "Type Id": "Family2: Family2",
                        "Phase Created": 1,
                        "Type Name": "Family2",
                        "Family Name": "Family2"
                    }
                }
            ],
            "account": "charence",
            "model": "6275729d-d772-4e21-a60f-eb6eb59360de"
        }
    ]
}
```

**Comment by mosment3DRepo** [comment]
@carmenfan looks good to me

**Review by carmenfan — APPROVED** [review]

**Pull request merged**